SAML Configuration

brand-fabian opened the pull request:
Basic adjustments to enable SAML authentication in VarFish. Most parts of the code are copied from the respective sodar-core files.

Note that for a working SAML authentication, it is very likely that users have to change the `ATTRIBUTES_MAP` fields in the saml config for the specific names. Therefore these fields should most likely be configurable through `env` as well.

holtgrewe commented:
@brand-fabian build failure is a known timing issue in CI here, will restart.

holtgrewe commented:
Please also add documentation to the manual on how to configure, maybe link to sodar-core docs where appropriate.

holtgrewe commented:
Also, please add change to HISTORY.

brand-fabian commented:
Currrently, we are still missing a way to configure the different values of the `ATTRIBUTES_MAP` field (as I posted in my original comment to #145). This is likely critical, since IDP's other than keycloak are excluded if they use other field names/field ids.

holtgrewe commented:
@stolpeo @mikkonie I think you need to take this over.

What we need to decide:

- do we limit ourselves to Keycloak only?
- if not, we need a way to specify the mapping in environment variables (e.g., JSON serialized Python dict)

Then we need to make this work for VarFish but also in sodar-cor or even SODAr.

I don't know what the best way forward is. Maybe we could make it work for @brand-fabian now (Keycloak, VarFish only) and create appropriate tickets for sodar-core and SODAR. Once we generalize the support and documentation we could push it to Varfish.

holtgrewe commented:
@brand-fabian internal discussion yields that you should be able to make `ATTRIBUTES_MAP` come through `env.dict`. Would that work here?

mikkonie commented:
Indeed, you should be able to do this by editing `base.py` as follows:

```
SAML2_AUTH = {
    # ...
    'ATTRIBUTES_MAP': env.dict(
        'SAML_ATTRIBUTES_MAP',
        {
            'email': 'Email',
            'username': 'UserName',
            'first_name': 'FirstName',
            'last_name': 'LastName',
        },
    )
}    
```
.. and then modify the value by providing a dict to `SAML_ATTRIBUTES_MAP` env var.

The only thing I'm not sure about is what's the best thing to have there as the default value. @stolpeo, any ideas?

brand-fabian commented:
I just added these env options to the base.py. Please take a look and check whether they are suitable. I have included my keycloak-based attributes map as default, but obviousy thats open to debate. If there are many use cases for Shibolleth or other SAML2 IDP's, then I can change it again.

Additionally, I moved over the login page changes from the corresponding Sodar-core file, to be able to have the SSO button on the login page.

holtgrewe commented:
Thanks, I'll get @stolpeo and @mikkonie check the parts that I know little of.

holtgrewe commented:
The CI fails with 

```
django.core.exceptions.ImproperlyConfigured: Set the SAML_NEW_USER_GROUPS environment variable
```

Please use defaults for all introduced settings.

mikkonie commented:
Hi @brand-fabian, one note about the login template. It seems you have copied it from django-sodar-core v0.10.x, while VarFish is still on v0.9.x. That means at least the icons will not work as expected. Until VarFish is bumped to the current SODAR Core version, I recommend copying the template changes [from here](https://cubi-gitlab.bihealth.org/CUBI_Engineering/CUBI_Data_Mgmt/sodar_core/-/blob/v0.9.1/projectroles/templates/projectroles/login.html) instead.
